### ISSUE TITLE
Shared subscription doesn't work well with partitioned topic in 2.4.2

### ISSUE THREAD

**Issue by Lanayx** [apache/pulsar#5877]
**Describe the bug**
Shared subscription doesn't work well with partitioned topic in 2.4.2. It worked well with 2.4.1 release. To reproduce issue I use 2 consumers with shared subscription and a topic with 2 partitions. Most of the times I run the test messages from only one partition are successfully consumed, while only the first message from another partition is consumed.

**To Reproduce**
1. Install 2.4.2 cluster (I used helm)
2. Setup topic with 2 partitions
3. Consume topic with 2 consumers in parallel that have shared subscription
4. Produce 50 messages in two partitions (with round robin or with 2 producers)
5. Observe issue

**Expected behavior**
All messages are consumed

**Screenshots**
![image](https://user-images.githubusercontent.com/3329606/70981925-2506ba00-20c7-11ea-911f-7237b096d3c1.png)

**Desktop (please complete the following information):**
 - Helm chart on minikube
 - Reproduced with dotnet client as well as java 2.4.1 and 2.4.2 clients

**Additional context**
The test that always worked on 2.4.1 now fails with 2.4.2 
https://github.com/fsharplang-ru/pulsar-client-dotnet/blob/develop/tests/IntegrationTests/Partitions.fs#L65-L150

**Comment by codelipenghui** [comment]
@Lanayx Could you help to check is your test still fail with master branch?

**Comment by Lanayx** [comment]
@codelipenghui I don't think so, I'm on windows, so only able to run pulsar as docker or kubernetes image. This is java code that I used to test those consumers if it helps (sorry for quality, I'm newbie in Java). 
```
package jtest;

import java.io.IOException;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.function.Function;
import java.util.logging.LogManager;

import org.apache.pulsar.client.api.Consumer;
import org.apache.pulsar.client.api.Producer;
import org.apache.pulsar.client.api.Message;
import org.apache.pulsar.client.api.MessageId;
import org.apache.pulsar.client.api.PulsarClient;
import org.apache.pulsar.client.api.PulsarClientException;
import org.apache.pulsar.client.api.Reader;
import org.apache.pulsar.client.api.SubscriptionInitialPosition;
import org.apache.pulsar.client.api.SubscriptionType;
import org.apache.pulsar.client.impl.TopicMessageIdImpl;

/**
 * Hello world!
 */
public final class App {
    private static Instant start;

    private App() {
    }

    public static void run(Consumer consumer, int number) {

        // Wait for a message
        CompletableFuture<Message> msgF = consumer.receiveAsync();
        msgF.thenAccept(msg -> {
            // try {
            // Do something with the message
            TopicMessageIdImpl id = (TopicMessageIdImpl) msg.getMessageId();
            System.out.printf("Message received: %s %s %s\n", msg.getKey(), new String(msg.getData()), id.getTopicPartitionName());

            // Acknowledge the message so that it can be deleted by the message broker
            consumer.acknowledgeAsync(msg).thenAccept(action -> {
                if (number < 5000) {
                    run(consumer, number + 1);
                } else {
                    Instant end = Instant.now();
                    long time = Duration.between(start, end).toMillis();
                    System.out.println("End! " + time);
                }

            }).exceptionally(x -> {
                System.out.println("Exception!");
                return x;
            });
        });

    }

    /**
     * Says hello to the world.
     *
     * @param args The arguments of the program.
     * @throws IOException
     */
    public static void main(String[] args) throws IOException {

        System.out.println("Hello World!");
        try {
            PulsarClient client = PulsarClient.builder().serviceUrl("pulsar://my-pulsar-cluster:31002").build();
            Consumer consumer = client.newConsumer()
                .topic("persistent://public/default/partitioned2")
                .subscriptionName("test-subscription3")
                .subscriptionInitialPosition(SubscriptionInitialPosition.Latest)
                .subscriptionType(SubscriptionType.Shared)
                .subscribe();

            Consumer consumer2 = client.newConsumer()
                .topic("persistent://public/default/partitioned2")
                .subscriptionName("test-subscription3")
                .subscriptionInitialPosition(SubscriptionInitialPosition.Latest)
                .subscriptionType(SubscriptionType.Shared)
                .subscribe();

            System.out.println("Consumer created! ");

            run(consumer,0);
            run(consumer2,0);

		} catch (PulsarClientException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
    }
}
```

**Comment by jiazhai** [comment]
@codelipenghui would you please help on this issue?

**Comment by codelipenghui** [comment]
@jiazhai I have test on 2.4.2 but can’t reproduce. @Lanayx I think you can disable batch, maybe messages are in a batch.

**Comment by Lanayx** [comment]
@codelipenghui I've disabled the batch, but see exactly the same behaviour. Have you tried reproducing several times on the same topic/subscription? I've just checked without batches and it worked once, on another run I see the same behavior.

**Comment by codelipenghui** [comment]
I have test by using your java code.

1.start 2 consumers
2.send some messages by using pulsar-client produce(50 messages)

But consumers can consume all messages:
```
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-1
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-0
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-1
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-1
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-1
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-0
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-0
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-0
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-0
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-0
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-0
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-0
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-0
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-0
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-1
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-0
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-0
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-0
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-0
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-1
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-1
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-0
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-0
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-0
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-1
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-1
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-1
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-1
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-0
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-0
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-0
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-0
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-1
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-1
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-0
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-1
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-1
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-1
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-1
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-1
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-0
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-0
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-0
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-0
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-1
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-1
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-1
[c42db]Message received:  {} persistent://public/default/partitioned2-partition-1
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-1
[993f2]Message received:  {} persistent://public/default/partitioned2-partition-1
```

I add consumer name before each log.

I think you can save the topic stats, stats-internal, heap dump of broker which owned that topic and heap dump of the client. This will facilitate troubleshooting.

**Comment by Lanayx** [comment]
Here are some stats, it's visible here 24 messages in backlog, and connected consumers, but those messages are not delivered until consumers reconnect
![image](https://user-images.githubusercontent.com/3329606/71067851-60b48900-2186-11ea-8dd0-bfbfd77c5c32.png)
![image](https://user-images.githubusercontent.com/3329606/71068174-36170000-2187-11ea-8491-bd36dacad4be.png)
![image](https://user-images.githubusercontent.com/3329606/71068809-8cd10980-2188-11ea-97a7-67480dde06e9.png)

**Comment by Lanayx** [comment]
Broker's logs
![image](https://user-images.githubusercontent.com/3329606/71069229-55169180-2189-11ea-9de9-5f1ea08db7c4.png)
![image](https://user-images.githubusercontent.com/3329606/71069427-c22a2700-2189-11ea-83b9-522dc5e04793.png)

**Comment by Lanayx** [comment]
Looks like the problem is not on consumer side, since the same behavior is seen on Java client 2.4.1, 2.4.2 and .net client that I'm developing.

**Comment by codelipenghui** [comment]
I have also recently checked a problem with the consumer stopping consumption. If you can easily reproduce, could you help to make a broker log(in debug level) and heap dump of broker(which owned the stop consume topic). Other users have also found this problem, but difficult to reproduce.

BTW, currently we used 2.4.0 , seems the problem has not happened yet.

**Comment by Lanayx** [comment]
Probably cursors are from other subscriptions, I've cut them for brevity. I don't think I can change debug level for broker, but I can take the dump, can you help me with that? Can't find a tool to see the processes on the node, top and ps are unavailable

**Comment by Lanayx** [comment]
Attaching the dump, hopefully I managed to take it correctly
[heap.zip](https://github.com/apache/pulsar/files/3978257/heap.zip)

**Comment by codelipenghui** [comment]
@Lanayx Maybe this issue related to 2.4.2 broker while dispatcher read 0 message from bookie, the dispatcher stop dispatch more messages. If your broker log contains following log, the issue will happens:

23:18:27.125 [bookkeeper-ml-workers-OrderedExecutor-0-0] DEBUG org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers - [persistent://message-platform/client-metrics/monitor-metrics / metrics-report] Distributing 0 messages to 4 consumers

keywords is ”Distributing 0 messages”

**Comment by Lanayx** [comment]
@codelipenghui can you please help with setting broker's log mode to debug? Helm template uses apachepulsar/pulsar-all docker image, I don't understand how to configure log level there.

**Comment by codelipenghui** [comment]
@Lanayx Maybe this issue related to #5894.

**Comment by Lanayx** [comment]
Fixed in 2.5.0
